### PR TITLE
ci/nightly: Compare parsec and parsec-tool dependency versions

### PR DESCRIPTION
**Pull request by tgonzalezorlandoarm**
Depends on #115 . Please only review when this one is merged.

- Compare dependencies between parsec and parsec-tool to spot common dependencies that are being used with different versions.
- Incorporate a nightly check to test this

Closes https://github.com/parallaxsecond/parsec/issues/360

### CODE REVIEW

**Comment by tgonzalezorlandoarm** [comment]
Triggered Nightly check:
https://github.com/tgonzalezorlandoarm/parsec-tool/actions/runs/6972003989

Note: This one should fail as parsec is currently using a different version of rustix

**Comment by tgonzalezorlandoarm** [comment]
Nightly run: https://github.com/tgonzalezorlandoarm/parsec-tool/actions/runs/6982621597

**Comment by tgonzalezorlandoarm** [comment]
Passing Nightly: https://github.com/tgonzalezorlandoarm/parsec-tool/actions/runs/6982894846/job/19002990252

File: utils/dependency_cross_matcher.py [hunk 1]
```diff
@@ -43,14 +43,31 @@ def main(argv=[], prog_name=''):
...
     parser.add_argument('--deps_dir',
                         required=True,
-                        help='Existing directory that contains the Cargo.toml for analyzing'
+                        nargs='+',
+                        help='Existing directories that contain Cargo.toml for analyzing '
                              'dependencies')
     args = parser.parse_args()
 
-    mismatches = run_deps_mismatcher(run_cargo_tree(args.deps_dir))
-    print_deps(mismatches)
```
Comment (gowthamsk-arm) [inline utils/dependency_cross_matcher.py:53]: At the beginning of the script in comments can you briefly explain what the script is intended to do and the ways it can be used?

File: utils/dependency_cross_matcher.py [hunk 2]
```diff
@@ -43,14 +43,31 @@ def main(argv=[], prog_name=''):
...
+    mismatches = dict()
+
+    if args.compare:
+        parsec_repo, parsec_tool_repo = args.deps_dir
+        mismatches_parsec = run_deps_mismatcher(run_cargo_tree(parsec_repo))
+        mismatches_parsec_tool = run_deps_mismatcher(run_cargo_tree(parsec_tool_repo))
```
Comment (gowthamsk-arm) [inline utils/dependency_cross_matcher.py:60]: I think the parsec service needs a few additional flags in the cargo tree command right?

File: utils/dependency_cross_matcher.py (hunk 1 above)
Comment (tgonzalezorlandoarm) [reply utils/dependency_cross_matcher.py:53]: This is done in here : https://github.com/parallaxsecond/parsec-tool/pull/116/commits/8defabb54df91a4458b67a2ab0312d9fa500b138#diff-f920291f1970d0b309680ec2e632ad0470e70bf983f097a1b40cc88c066fbb87R43

Should I rephrase it to be clearer or do you want me to duplicate the comment?

This information can also be accessed by calling:
```
python3 utils/dependency_cross_matcher.py --help
usage: DependencyCrossmatcher [-h] [-c] --deps_dir DEPS_DIR [DEPS_DIR ...]

Checks the version mismatches for dependencies in Cargo based repositories

options:
  -h, --help            show this help message and exit
  -c, --compare         Check for mismatches between 2 repositories
  --deps_dir DEPS_DIR [DEPS_DIR ...]
                        Existing directories that contain Cargo.toml for analyzing dependencies
```

Comment (gowthamsk-arm) [reply utils/dependency_cross_matcher.py:53]: Yeah, this test block if you can add them as a comment section in the beginning I feel it would be easier for people to understand it rather than going through the code. What do you think?

File: utils/dependency_cross_matcher.py (hunk 2 above)
Comment (tgonzalezorlandoarm) [reply utils/dependency_cross_matcher.py]: Done

**Comment by tgonzalezorlandoarm** [comment]
https://github.com/tgonzalezorlandoarm/parsec-tool/actions/runs/6983462680

File: utils/dependency_cross_matcher.py (hunk 2 above)
Comment (tgonzalezorlandoarm) [reply utils/dependency_cross_matcher.py]: Thanks for noticing! Well spotted

File: utils/dependency_cross_matcher.py (hunk 1 above)
Comment (tgonzalezorlandoarm) [reply utils/dependency_cross_matcher.py:53]: Done. I'm not retriggering the CI Nightly run (that passed) as it's only a comment

File: tests/ci.sh [hunk 3]
```diff
@@ -34,11 +34,10 @@ done
...
 #########################
 if [ "$MISMATCHER" = "True" ]; then
+
     python3 $(pwd)/utils/dependency_cross_matcher.py --deps_dir $(pwd)
-    mismatcher_result=$?
-    if [ "$mismatcher_result" -ne 0 ]; then
-        error_msg "Found dependencies version mismatches"
-    fi
+
+	python3 $(pwd)/utils/dependency_cross_matcher.py -c --deps_dir $(pwd)/parsec $(pwd)
```
Comment (gowthamsk-arm) [inline tests/ci.sh:40]: nit: spacing?

Comment (tgonzalezorlandoarm) [reply tests/ci.sh:40]: Sorry, what spacing? The added newline?
I thought it looked nicer that way?

Comment (tgonzalezorlandoarm) [reply tests/ci.sh:40]: Oh, I see, the tab. I don't see it in my IDE, and the script is working in the CI. THat's weird

Comment (tgonzalezorlandoarm) [reply tests/ci.sh:40]: Will fix it, it was a "tab" instead of spaces

**Comment by tgonzalezorlandoarm** [comment]
Once this is merged, I will create a PR in https://github.com/parallaxsecond/parsec to keep the tests synced.

**Review by gowthamsk-arm — APPROVED** [review]
